Runtime staging compiler crash when creating an object using "new" (scala 3.3.1)

## Compiler version

3.3.1

## Minimized code

import scala.quoted.*
given staging.Compiler = staging.Compiler.make(getClass.getClassLoader)

class A(i: Int)

def f(i: Expr[Int])(using Quotes): Expr[A] = { '{ new A($i) } }

val g: Int => A = staging.run { '{ (i: Int) => ${ f('{i}) } } }



## Output

scala> g
dotty.tools.dotc.reporting.UnhandledError: undefined: new A # -1: TermRef(TypeRef(ThisType(TypeRef(ThisType(TypeRef(NoPrefix,module class <root>)),module class <empty>)),A),<init>) at quotedFrontend
  at dotty.tools.dotc.reporting.ThrowingReporter.doReport(ThrowingReporter.scala:14)
  at dotty.tools.dotc.reporting.Reporter.issueUnconfigured(Reporter.scala:158)
  at dotty.tools.dotc.reporting.Reporter.go$1(Reporter.scala:181)
  at dotty.tools.dotc.reporting.Reporter.issueIfNotSuppressed(Reporter.scala:200)
  at dotty.tools.dotc.reporting.Reporter.report(Reporter.scala:203)
  at dotty.tools.dotc.report$.error(report.scala:68)
  at dotty.tools.dotc.typer.ErrorReporting$.errorType(ErrorReporting.scala:31)
  at dotty.tools.dotc.typer.TypeAssigner.assignType(TypeAssigner.scala:298)
  at dotty.tools.dotc.typer.TypeAssigner.assignType$(TypeAssigner.scala:16)
  at dotty.tools.dotc.typer.Typer.assignType(Typer.scala:116)
  at dotty.tools.dotc.ast.tpd$.Apply(tpd.scala:49)
  at dotty.tools.dotc.core.tasty.TreeUnpickler$TreeReader.readLengthTree$1(TreeUnpickler.scala:1304)
  at dotty.tools.dotc.core.tasty.TreeUnpickler$TreeReader.readTree(TreeUnpickler.scala:1468)
  at dotty.tools.dotc.core.tasty.TreeUnpickler$TreeReader.readLengthTree$1(TreeUnpickler.scala:1333)
  at dotty.tools.dotc.core.tasty.TreeUnpickler$TreeReader.readTree(TreeUnpickler.scala:1468)
  at dotty.tools.dotc.core.tasty.TreeUnpickler.unpickle(TreeUnpickler.scala:117)
  at dotty.tools.dotc.core.tasty.DottyUnpickler.computeRootTrees(DottyUnpickler.scala:62)
  at dotty.tools.dotc.ast.tpd$TreeProvider.rootTrees(tpd.scala:1299)
  at dotty.tools.dotc.ast.tpd$TreeProvider.rootTrees$(tpd.scala:1288)
  at dotty.tools.dotc.core.tasty.DottyUnpickler.rootTrees(DottyUnpickler.scala:44)
  at dotty.tools.dotc.ast.tpd$TreeProvider.tree(tpd.scala:1303)
  at dotty.tools.dotc.ast.tpd$TreeProvider.tree$(tpd.scala:1288)
  at dotty.tools.dotc.core.tasty.DottyUnpickler.tree(DottyUnpickler.scala:44)
  at dotty.tools.dotc.quoted.PickledQuotes$.unpickle(PickledQuotes.scala:274)
  at dotty.tools.dotc.quoted.PickledQuotes$.unpickleTerm(PickledQuotes.scala:84)
  at scala.quoted.runtime.impl.QuotesImpl.unpickleExprV2(QuotesImpl.scala:3143)
  at Meta$package$.f(Meta.scala:7)
  at Meta$package$.$init$$$anonfun$1$$anonfun$1$$anonfun$1(Meta.scala:9)
  at Meta$package$.$init$$$anonfun$1$$anonfun$1(Meta.scala:9)
  at Meta$package$.$init$$$anonfun$1$$anonfun$adapted$1(Meta.scala:9)
  at dotty.tools.dotc.quoted.PickledQuotes$$anon$1.transform(PickledQuotes.scala:110)
  at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform(Trees.scala:1538)
  at dotty.tools.dotc.quoted.PickledQuotes$$anon$1.transform(PickledQuotes.scala:135)
  at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transformBlock(Trees.scala:1603)
  at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform(Trees.scala:1518)
  at dotty.tools.dotc.quoted.PickledQuotes$$anon$1.transform(PickledQuotes.scala:135)
  at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform(Trees.scala:1570)
  at dotty.tools.dotc.quoted.PickledQuotes$$anon$1.transform(PickledQuotes.scala:135)
  at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform$$anonfun$1(Trees.scala:1605)
  at scala.collection.immutable.List.mapConserve(List.scala:472)
  at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform(Trees.scala:1605)
  at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transformStats(Trees.scala:1601)
  at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transformBlock(Trees.scala:1603)
  at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform(Trees.scala:1518)
  at dotty.tools.dotc.quoted.PickledQuotes$$anon$1.transform(PickledQuotes.scala:135)
  at dotty.tools.dotc.quoted.PickledQuotes$.spliceTerms(PickledQuotes.scala:152)
  at dotty.tools.dotc.quoted.PickledQuotes$.unpickleTerm(PickledQuotes.scala:88)
  at scala.quoted.runtime.impl.QuotesImpl.unpickleExprV2(QuotesImpl.scala:3143)
  at Meta$package$.$init$$$anonfun$1(Meta.scala:9)
  at scala.quoted.staging.QuoteCompiler$QuotedFrontend.runOn$$anonfun$1(QuoteCompiler.scala:83)
  at scala.collection.immutable.List.flatMap(List.scala:293)
  at scala.quoted.staging.QuoteCompiler$QuotedFrontend.runOn(QuoteCompiler.scala:98)
  at dotty.tools.dotc.Run.runPhases$1$$anonfun$1(Run.scala:246)
  at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
  at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
  at scala.collection.ArrayOps$.foreach$extension(ArrayOps.scala:1321)
  at dotty.tools.dotc.Run.runPhases$1(Run.scala:262)
  at dotty.tools.dotc.Run.compileUnits$$anonfun$1(Run.scala:270)
  at dotty.tools.dotc.Run.compileUnits$$anonfun$adapted$1(Run.scala:279)
  at dotty.tools.dotc.util.Stats$.maybeMonitored(Stats.scala:67)
  at dotty.tools.dotc.Run.compileUnits(Run.scala:279)
  at dotty.tools.dotc.Run.compileUnits(Run.scala:200)
  at scala.quoted.staging.QuoteCompiler$ExprRun.compileExpr(QuoteCompiler.scala:118)
  at scala.quoted.staging.QuoteDriver.run(QuoteDriver.scala:43)
  at scala.quoted.staging.Compiler$$anon$1.run(Compiler.scala:38)
  at scala.quoted.staging.package$.run(staging.scala:19)
  at Meta$package$.<clinit>(Meta.scala:9)
  ... 66 elided


## Expectation

...for this not to happen.

Compiling the file below will emit a valid warning 
```scala
import scala.quoted.*
given staging.Compiler = staging.Compiler.make(getClass.getClassLoader)
class A(i: Int)
def f(i: Expr[Int])(using Quotes): Expr[A] = { '{ new A($i) } }
val g: Int => A = staging.run { '{ (i: Int) => ${ f('{i}) } } }
@main def run = g
```
Full stack trace : 
```shell
-- [E181] Potential Issue Warning: /Users/hamzaremmal/Desktop/LAMP/dotty/i19170.scala:3:47
3 |given staging.Compiler = staging.Compiler.make(getClass.getClassLoader)
  |                                               ^^^^^^^^
  |                         Suspicious top-level unqualified call to getClass
  |
  | longer explanation available when compiling with `-explain`
1 warning found
java.lang.ExceptionInInitializerError
	at run.main(i19170.scala:11)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at dotty.tools.scripting.ScriptingDriver.compileAndRun(ScriptingDriver.scala:33)
	at dotty.tools.scripting.Main$.process(Main.scala:45)
	at dotty.tools.MainGenericRunner$.run$1(MainGenericRunner.scala:250)
	at dotty.tools.MainGenericRunner$.process(MainGenericRunner.scala:270)
	at dotty.tools.MainGenericRunner$.main(MainGenericRunner.scala:281)
	at dotty.tools.MainGenericRunner.main(MainGenericRunner.scala)
Caused by: dotty.tools.dotc.reporting.UnhandledError: undefined: new A # -1: TermRef(TypeRef(ThisType(TypeRef(ThisType(TypeRef(NoPrefix,module class <root>)),module class <empty>)),A),<init>) at quotedFrontend
	at dotty.tools.dotc.reporting.ThrowingReporter.doReport(ThrowingReporter.scala:14)
	at dotty.tools.dotc.reporting.Reporter.issueUnconfigured(Reporter.scala:158)
	at dotty.tools.dotc.reporting.Reporter.go$1(Reporter.scala:181)
	at dotty.tools.dotc.reporting.Reporter.issueIfNotSuppressed(Reporter.scala:200)
	at dotty.tools.dotc.reporting.Reporter.report(Reporter.scala:203)
	at dotty.tools.dotc.report$.error(report.scala:68)
	at dotty.tools.dotc.typer.ErrorReporting$.errorType(ErrorReporting.scala:31)
	at dotty.tools.dotc.typer.TypeAssigner.assignType(TypeAssigner.scala:298)
	at dotty.tools.dotc.typer.TypeAssigner.assignType$(TypeAssigner.scala:16)
	at dotty.tools.dotc.typer.Typer.assignType(Typer.scala:116)
	at dotty.tools.dotc.ast.tpd$.Apply(tpd.scala:49)
	at dotty.tools.dotc.core.tasty.TreeUnpickler$TreeReader.readLengthTree$1(TreeUnpickler.scala:1304)
	at dotty.tools.dotc.core.tasty.TreeUnpickler$TreeReader.readTree(TreeUnpickler.scala:1468)
	at dotty.tools.dotc.core.tasty.TreeUnpickler$TreeReader.readLengthTree$1(TreeUnpickler.scala:1333)
	at dotty.tools.dotc.core.tasty.TreeUnpickler$TreeReader.readTree(TreeUnpickler.scala:1468)
	at dotty.tools.dotc.core.tasty.TreeUnpickler.unpickle(TreeUnpickler.scala:117)
	at dotty.tools.dotc.core.tasty.DottyUnpickler.computeRootTrees(DottyUnpickler.scala:62)
	at dotty.tools.dotc.ast.tpd$TreeProvider.rootTrees(tpd.scala:1299)
	at dotty.tools.dotc.ast.tpd$TreeProvider.rootTrees$(tpd.scala:1288)
	at dotty.tools.dotc.core.tasty.DottyUnpickler.rootTrees(DottyUnpickler.scala:44)
	at dotty.tools.dotc.ast.tpd$TreeProvider.tree(tpd.scala:1303)
	at dotty.tools.dotc.ast.tpd$TreeProvider.tree$(tpd.scala:1288)
	at dotty.tools.dotc.core.tasty.DottyUnpickler.tree(DottyUnpickler.scala:44)
	at dotty.tools.dotc.quoted.PickledQuotes$.unpickle(PickledQuotes.scala:274)
	at dotty.tools.dotc.quoted.PickledQuotes$.unpickleTerm(PickledQuotes.scala:84)
	at scala.quoted.runtime.impl.QuotesImpl.unpickleExprV2(QuotesImpl.scala:3143)
	at i19170$package$.f(i19170.scala:7)
	at i19170$package$.$init$$$anonfun$1$$anonfun$1$$anonfun$1(i19170.scala:9)
	at i19170$package$.$init$$$anonfun$1$$anonfun$1(i19170.scala:9)
	at i19170$package$.$init$$$anonfun$1$$anonfun$adapted$1(i19170.scala:9)
	at dotty.tools.dotc.quoted.PickledQuotes$$anon$1.transform(PickledQuotes.scala:110)
	at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform(Trees.scala:1538)
	at dotty.tools.dotc.quoted.PickledQuotes$$anon$1.transform(PickledQuotes.scala:135)
	at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transformBlock(Trees.scala:1603)
	at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform(Trees.scala:1518)
	at dotty.tools.dotc.quoted.PickledQuotes$$anon$1.transform(PickledQuotes.scala:135)
	at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform(Trees.scala:1570)
	at dotty.tools.dotc.quoted.PickledQuotes$$anon$1.transform(PickledQuotes.scala:135)
	at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform$$anonfun$1(Trees.scala:1605)
	at scala.collection.immutable.List.mapConserve(List.scala:472)
	at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform(Trees.scala:1605)
	at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transformStats(Trees.scala:1601)
	at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transformBlock(Trees.scala:1603)
	at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform(Trees.scala:1518)
	at dotty.tools.dotc.quoted.PickledQuotes$$anon$1.transform(PickledQuotes.scala:135)
	at dotty.tools.dotc.quoted.PickledQuotes$.spliceTerms(PickledQuotes.scala:152)
	at dotty.tools.dotc.quoted.PickledQuotes$.unpickleTerm(PickledQuotes.scala:88)
	at scala.quoted.runtime.impl.QuotesImpl.unpickleExprV2(QuotesImpl.scala:3143)
	at i19170$package$.$init$$$anonfun$1(i19170.scala:9)
	at scala.quoted.staging.QuoteCompiler$QuotedFrontend.runOn$$anonfun$1(QuoteCompiler.scala:83)
	at scala.collection.immutable.List.flatMap(List.scala:293)
	at scala.quoted.staging.QuoteCompiler$QuotedFrontend.runOn(QuoteCompiler.scala:98)
	at dotty.tools.dotc.Run.runPhases$1$$anonfun$1(Run.scala:246)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.ArrayOps$.foreach$extension(ArrayOps.scala:1321)
	at dotty.tools.dotc.Run.runPhases$1(Run.scala:262)
	at dotty.tools.dotc.Run.compileUnits$$anonfun$1(Run.scala:270)
	at dotty.tools.dotc.Run.compileUnits$$anonfun$adapted$1(Run.scala:279)
	at dotty.tools.dotc.util.Stats$.maybeMonitored(Stats.scala:67)
	at dotty.tools.dotc.Run.compileUnits(Run.scala:279)
	at dotty.tools.dotc.Run.compileUnits(Run.scala:200)
	at scala.quoted.staging.QuoteCompiler$ExprRun.compileExpr(QuoteCompiler.scala:118)
	at scala.quoted.staging.QuoteDriver.run(QuoteDriver.scala:43)
	at scala.quoted.staging.Compiler$$anon$1.run(Compiler.scala:38)
	at scala.quoted.staging.package$.run(staging.scala:19)
	at i19170$package$.<clinit>(i19170.scala:9)
	... 11 more
```

--- a/compiler/src/dotty/tools/dotc/Run.scala
+++ b/compiler/src/dotty/tools/dotc/Run.scala
@@ -661,4 +661,6 @@ object Run {
         report.enrichErrorMessage(errorMessage)
       else
         errorMessage
+    def doNotEnrichErrorMessage: Unit =
+      if run != null then run.myEnrichedErrorMessage = true
 }

--- a/staging/src/scala/quoted/staging/QuoteCompiler.scala
+++ b/staging/src/scala/quoted/staging/QuoteCompiler.scala
@@ -48,7 +48,9 @@ private class QuoteCompiler extends Compiler:
 
   override def newRun(implicit ctx: Context): ExprRun =
     reset()
-    new ExprRun(this, ctx.addMode(Mode.ReadPositions))
+    val run = new ExprRun(this, ctx.addMode(Mode.ReadPositions))
+    run.doNotEnrichErrorMessage
+    run
 
   def outputClassName: TypeName = "Generated$Code$From$Quoted".toTypeName
 

--- a/staging/src/scala/quoted/staging/QuoteDriver.scala
+++ b/staging/src/scala/quoted/staging/QuoteDriver.scala
@@ -8,6 +8,7 @@ import dotty.tools.dotc.quoted.QuotesCache
 import dotty.tools.io.{AbstractFile, Directory, PlainDirectory, VirtualDirectory}
 import dotty.tools.repl.AbstractFileClassLoader
 import dotty.tools.dotc.reporting._
+import dotty.tools.dotc.config.Settings.Setting.value
 import dotty.tools.dotc.util.ClasspathFromClassloader
 import scala.quoted._
 import scala.quoted.staging.Compiler
@@ -40,7 +41,20 @@ private class QuoteDriver(appClassloader: ClassLoader) extends Driver:
       setCompilerSettings(ctx1.fresh.setSetting(ctx1.settings.outputDir, outDir), settings)
     }
 
-    new QuoteCompiler().newRun(ctx).compileExpr(exprBuilder) match
+    val compiledExpr = 
+      try
+        new QuoteCompiler().newRun(ctx).compileExpr(exprBuilder)
+      catch case ex: dotty.tools.FatalError =>
+        val enrichedMessage =
+          s"""An unhandled exception was thrown in the staging compiler.
+            |This might be caused by using an incorrect classloader
+            |when creating the `staging.Compiler` instance with `staging.Compiler.make`.
+            |For details, please refer to the documentation.
+            |For non-enriched exceptions, compile with -Yno-enrich-error-messages.""".stripMargin
+        if ctx.settings.YnoEnrichErrorMessages.value(using ctx) then throw ex
+        else throw new Exception(enrichedMessage, ex)
+    
+    compiledExpr match
       case Right(value) =>
         value.asInstanceOf[T]
 

--- a/tests/run-staging/i19170c.check
+++ b/tests/run-staging/i19170c.check
@@ -1,0 +1,1 @@
+exception thrown, no additional printlns

--- a/tests/run-staging/i19170c.scala
+++ b/tests/run-staging/i19170c.scala
@@ -1,0 +1,16 @@
+import scala.quoted.*
+
+given staging.Compiler =
+  staging.Compiler.make(getClass.getClassLoader.getParent) // different classloader that 19170b.scala
+class A(i: Int)
+
+def f(i: Expr[Int])(using Quotes): Expr[A] = { '{ new A($i) } }
+
+@main def Test = {
+  try
+    val g: Int => A = staging.run { '{ (i: Int) => ${ f('{i}) } } }
+    println(g(3))
+  catch case ex: Exception =>
+    assert(ex.getMessage().startsWith("An unhandled exception was thrown in the staging compiler."), ex.getMessage())
+    println("exception thrown, no additional printlns")
+}

--- a/tests/run-staging/i19176b.check
+++ b/tests/run-staging/i19176b.check
@@ -1,0 +1,1 @@
+exception thrown, no additional printlns

--- a/tests/run-staging/i19176b.scala
+++ b/tests/run-staging/i19176b.scala
@@ -1,0 +1,14 @@
+import scala.quoted.*
+
+given staging.Compiler =
+  staging.Compiler.make(getClass.getClassLoader.getParent) // we want to make sure the classloader is incorrect
+
+class A
+
+@main def Test =
+  try
+    val f: (A, Int) => Int = staging.run { '{ (q: A, x: Int) => x } }
+    f(new A, 3)
+  catch case ex: Exception =>
+    assert(ex.getMessage().startsWith("An unhandled exception was thrown in the staging compiler."), ex.getMessage())
+    println("exception thrown, no additional printlns")

